src/csharp/Microsoft.Spark/Broadcast.cs

```diff
@@ -0,0 +1,243 @@
+        private readonly string _path;
+
+        internal Broadcast(SparkContext sc,
+            T value,
+            string localDir,
+            JvmObjectReference sparkContext)
+        {
+            var tempDir = Path.Combine(localDir, "sparkdotnet");
+            _path = Path.Combine(tempDir, Path.GetRandomFileName());
+            Version version = SparkEnvironment.SparkVersion;
+
+            _jvmObject = (version.Major, version.Minor) switch
+            {
+                (2, 3) when version.Build == 0 || version.Build == 1 =>
+                    CreateBroadcast_V2_3_1_Below(tempDir, sparkContext, value),
+                (2, 3) => CreateBroadcast_V2_3_2_Above(sc, tempDir, value),
+                (2, 4) => CreateBroadcast_V2_3_2_Above(sc, tempDir, value),
+                _ => throw new NotSupportedException($"Spark {version} not supported.")
+            };
+
+            _bid = (long)_jvmObject.Invoke("id");
+            JvmBroadcastRegistry.Add(_jvmObject);
+        }
+
+        JvmObjectReference IJvmObjectReferenceProvider.Reference => _jvmObject;
+
+        /// <summary>
+        /// Get the broadcasted value.
+        /// </summary>
+        /// <returns>The broadcasted value</returns>
+        public T Value()
+        {
+            return (T)BroadcastRegistry.Get(_bid);
+        }
+
+        /// <summary>
+        /// Asynchronously delete cached copies of this broadcast on the executors.
+        /// If the broadcast is used after this is called, it will need to be re-sent to each
+        /// executor.
+        /// </summary>
+        public void Unpersist()
+        {
+            _jvmObject.Invoke("unpersist");
+        }
+
+        /// <summary>
+        /// Delete cached copies of this broadcast on the executors. If the broadcast is used after
+        /// this is called, it will need to be re-sent to each executor.
+        /// </summary>
+        /// <param name="blocking">Whether to block until unpersisting has completed</param>
+        public void Unpersist(bool blocking)
+        {
+            _jvmObject.Invoke("unpersist", blocking);
+        }
+
+        /// <summary>
+        /// Destroy all data and metadata related to this broadcast variable. Use this with
+        /// caution; once a broadcast variable has been destroyed, it cannot be used again.
+        /// This method blocks until destroy has completed.
+        /// </summary>
+        public void Destroy()
+        {
+            _jvmObject.Invoke("destroy");
+            File.Delete(_path);
+        }
+
+        /// <summary>
+        /// Function that creates a file to store the broadcast value in the given path.
+        /// </summary>
+        /// <param name="tempDir">Path where file is to be created</param>
+        /// <param name="value">Broadcast value to be written to the file</param>
+        private void WriteBroadcastValueToFile(string tempDir, object value)
+        {
+            Directory.CreateDirectory(tempDir);
```

Comment:
When pyspark creates this tempDir, it automatically gets cleaned up by the JVM when it shuts down.  Are we doing anything similar with this tempDir ?

Reply:
I see this tempDir get's cleaned up locally on my system once it shuts down

Reply:
Yes, the tempDir gets cleaned up automatically on shut down.

Reply:
Do you know what is cleaning it up ?

Reply:
Not sure, looking into it.

Reply:
oh Spark should be cleaning this up since you get the local directory from Spark. :)

Reply:
I see, I was confused because I saw Spark does not delete the localDir, just the tempDir gets deleted so was not sure what was happening. Thanks @imback82 ! Also do we know why localDir is also not cleaned up? Just curious.

Reply:
isn't that the windows bug that you get at the end of the run? (you can double check the path)

Reply:
Yes that is the correct, it is the same path. Thanks :)